src/python/pants/backend/adhoc/adhoc_tool_test.py

```diff
@@ -373,6 +371,7 @@ def test_adhoc_tool_workspace_invalidation_sources(rule_runner: PythonRuleRunner
 
     # Update the hash-only source file's content. The adhoc_tool should be re-executed now.
     (Path(rule_runner.build_root) / "src" / "a-file").write_text("xyzzy")
+    time.sleep(0.1)  # wait for invalidation to occur in engine
```

Comment:
This seems shady?

Reply:
The invalidation here only occurs once the filesystem watch logic invalidates for changes. For some reason, the invalidation was taking too long to occur on my system when testing.